src/sync/sync_permission.cpp

```diff
@@ -0,0 +1,218 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2017 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "sync/sync_permission.hpp"
+
+#include "impl/object_accessor_impl.hpp"
+#include "object_schema.hpp"
+#include "property.hpp"
+
+#include "sync/sync_config.hpp"
+#include "sync/sync_manager.hpp"
+#include "sync/sync_session.hpp"
+#include "sync/sync_user.hpp"
+#include "util/event_loop_signal.hpp"
+#include "util/uuid.hpp"
+
+#include <realm/query_expression.hpp>
+#include <realm/util/to_string.hpp>
+
+using namespace realm;
+
+// MARK: - Utility
+
+namespace {
+
+Permission::AccessLevel extract_access_level(Object& permission, CppContext& context)
+{
+    auto may_manage = permission.get_property_value<util::Any>(&context, "mayManage");
+    if (may_manage.has_value() && any_cast<bool>(may_manage))
+        return Permission::AccessLevel::Admin;
+
+    auto may_write = permission.get_property_value<util::Any>(&context, "mayWrite");
+    if (may_write.has_value() && any_cast<bool>(may_write))
+        return Permission::AccessLevel::Write;
+
+    auto may_read = permission.get_property_value<util::Any>(&context, "mayRead");
+    if (may_read.has_value() && any_cast<bool>(may_read))
+        return Permission::AccessLevel::Read;
+
+    return Permission::AccessLevel::None;
+}
+
+}
+
+// MARK: - PermissionResults
+
+Permission PermissionResults::get(size_t index)
+{
+    Object permission(m_results.get_realm(), m_results.get_object_schema(), m_results.get(index));
+    CppContext context;
+    return {
+        any_cast<std::string>(permission.get_property_value<util::Any>(&context, "path")),
+        extract_access_level(permission, context),
+        { any_cast<std::string>(permission.get_property_value<util::Any>(&context, "userId")) }
+    };
+}
+
+// MARK: - Permissions
+
+// A wrapper that stores a value and an associated notification token.
+// The point of this type is to keep the notification token alive
+// until the value can be properly processed or handled.
+template<typename T>
+struct NotificationWrapper {
```

Comment:
Thoughts on doing something like https://gist.github.com/bdash/ede60c67ac3ca65d43985b70dd8fdf7c to make this more ergonomic?